### PR TITLE
Adds lm_eval to evaluations

### LINKED ISSUE #199: Multi-Dataset Validation with Generative Benchmarks
# 🎯 **Goal (What & Why)**

Improve Fast-LLM’s validation pipeline by enabling multiple independent validation datasets, allowing better insights into model performance across different domains. This will streamline evaluations, reduce manual steps, and align with best practices in modern LLM training.

# 🚀 **Execution Plan**
### **Step: Introduce Generative Benchmarks (later)**

- Add support for generative evaluations, where:
  - The model is prompted using predefined queries taken from dataset.
  - Greedy decoding is used for inference.
  - A simple extractor function determines success/failure based on reference outputs.
- LM-eval-harness or similar evaluation frameworks could serve as the foundation of structured benchmarking.

# 📌 **Acceptance Criteria** (Must-Haves for Completion)

* Multi-dataset Generative Benchmarks evaluation must be implemented and validated.
* Results must be logged to WandB.

# 🛠️ **Project Management**
- [x] **Assign the project to the Fast-LLM project.**
- [ ] **Set the `Estimate` field (in days) in the GitHub project.**
- [ ] **Use the `Size` field to categorize the PR size (Small/Medium/Large).**
- [ ] **Assign an owner when opening the issue.**

@jlamypoirier  comments on this:

> * Generalize the validation scheme to allow for various model evaluations (i.e. make an abstrace base class similar to `Evaluator`).
> * Allow for multiple independent evaluations, each with their own config (including dataset, if applicable)
> * Implement evaluator(s) for generative benchmark according to our needs (wrap other tools?)
> * Perform basic optimizations for generative inference (ex. kv caching, flash-attn for generation)
> * Maybe a few mote things to make sure Fast-LLM works for generative inference.

### PR DESCRIPTION
**Pull request by bigximik**
# ✨ Description

Add **lm_eval** integration which enables running evaluation directly from an in-memory model. Currently, only data parallelism is supported.

Closes #199

## 🔍 Type of change

Select all that apply:

- [x] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [x] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. **Improved Weights & Biases (wandb) Logging**  
   The final log for each step now explicitly commits the data, ensuring that the step is immediately visible in the wandb interface. Previously, a step would only appear after logging began for the following step.

2. **Integration of `lm_eval` Evaluator**  
   Added support for `lm_eval` as a built-in evaluation method. This enables running standard language model benchmarks during or after training using the Evaluation Harness.

3. **Extended Support for Distributed Primitives**  
   Introduced custom distributed utilities adapted from PyTorch to support non-standard process groups to support lm_eval integration.

4. **Added User Guide for Evaluations**  
   Included documentation explaining how to configure and use evaluators such as `loss` and `lm_eval`, both during training and in standalone evaluation mode.

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [x] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/developers/contributing).
- [x] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [x] 🎉 The functionality is complete, and I have tested the changes.
- [x] 📝 I have updated the documentation if needed.
- [x] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [x] 🧩 I have commented my code, especially in hard-to-understand areas.

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [x] ✔️ New and existing tests pass locally with my changes.
- [x] 🚦 I have tested these changes on GPUs and verified training stability.
- [x] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### CODE REVIEW

**Comment by bigximik** [comment]
Log-likelihood calculation in `lm_eval` wrapper is slow because it is performed on the CPU, see TODOs and NOTEs in code. However, it is still faster than generation without KV cache. I propose optimizing it after KV cache implementation in a separate PR.

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,946 @@
...
+        # Messages could include types like logits, generate, finished.
+
+        if self.group is None or (world_size := self.group.size()) == 1:
+            # Must not be called with continue_generate false on one process
+            assert continue_generate
+            return self._model_invoke_inner(
+                input_ids, attention_mask, labels, max_length, stop, generate, **generation_kwargs
+            )
+
+        rank = self.group.rank()
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:119]: Unnecessary variable

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,946 @@
...
+                    obj_list,
+                    scatter_list,
+                    group_src=0,
+                    group=self.group,
+                )
+                input_ids, attention_mask, labels, max_length, stop, generate, continue_generate, generation_kwargs = (
+                    tuple(obj_list[0])
+                )
+
+                if continue_generate == False:
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:216]: `if not continue_generate`

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,946 @@
...
+            return self._model_generate_inner(input_ids, attention_mask, max_length, stop, **generation_kwargs)
+        else:
+            return self._model_call_inner(input_ids, attention_mask, labels)
+
+    def _model_call(self, input_ids, attention_mask=None, labels=None):
+        return self._model_invoke(
+            input_ids, attention_mask, labels, None, None, generate=False, continue_generate=True
+        )
+
+    def _model_generate(self, input_ids, attention_mask, max_length, stop, **generation_kwargs):
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:259]: These methods look duplicate, are they both necessary?

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,946 @@
...
+                    position_ids=None,
+                    past_key_values=None,
+                    inputs_embeds=None,
+                    use_cache=False,
+                    output_attentions=False,
+                    output_hidden_states=False,
+                    return_dict=True,
+                ).logits.cpu()
+            else:
+                return self.model(
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:312]: The two calls look identical

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,946 @@
...
+            generation_kwargs["do_sample"] = do_sample = False
+
+        if do_sample is False and generation_kwargs.get("temperature") == 0.0:
+            generation_kwargs.pop("temperature")
+        # build stopping criteria
+        stopping_criteria = lm_eval.models.utils.stop_sequences_criteria(
+            self.tokenizer, stop, input_ids.shape[1], input_ids.shape[0]
+        )
+        if attention_mask is None:
+            return self.model.generate(
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:344]: Calls also look identical

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,946 @@
...
+
+    @property
+    def config(self):
+        # return the associated transformers.AutoConfig for the given pretrained model.
+        return self._config
+
+    @property
+    def model(self):
+        # returns the model, unwrapping it if using Accelerate
+        if hasattr(self, "accelerator"):
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:371]: We don't need this

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,946 @@
...
+    @property
+    def max_length(self):
+        if self._max_length:  # if max length manually set, return it
+            return self._max_length
+        seqlen_config_attrs = ("n_positions", "max_position_embeddings", "n_ctx")
+        for attr in seqlen_config_attrs:
+            if hasattr(self.model.config, attr):
+                return getattr(self.model.config, attr)
+        if hasattr(self.tokenizer, "model_max_length"):
+            if self.tokenizer.model_max_length == 1000000000000000019884624838656:
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:399]: ?

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,946 @@
...
+    def prefix_token_id(self):
+        # it is used as prefix for loglikelihood
+        if self.custom_prefix_token_id is not None:
+            return self.custom_prefix_token_id
+        if self.tokenizer.bos_token_id is not None:
+            return self.tokenizer.bos_token_id
+        return self.tokenizer.eos_token_id
+
+    @property
+    def max_length(self):
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:391]: Can't we just get it from the model instead?

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,946 @@
...
+    @property
+    def max_gen_toks(self) -> int:
+        return 256
+
+    # TODO: check removing this does not affect lm_eval
+    # @property
+    # def batch_size(self):
+    #     return self.batch_size_per_gpu
+
+    @property
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:413]: Are all these properties needed?

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,946 @@
...
+    def _loglikelihood_tokens(
+        self,
+        requests: list[tuple[tuple[str, str], list[int], list[int]]],
+        disable_tqdm: bool = False,
+        override_bs: int = None,
+    ) -> list[tuple[float, bool]]:
+        # TODO: implement some kind of efficient-request-middleware that lumps together requests with the same context
+        res = []
+
+        def _collate(req: tuple[tuple[str, str], list[int], list[int]]):
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:609]: lambdas would be simpler here

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,946 @@
...
+        if (len(self.batch_sizes) > 1) and (self.batch_sizes[sched - 1] == self.max_batch_size):
+            # if previous batch size is already maximal, skip recomputation
+            self.batch_sizes[sched] = self.max_batch_size
+            return self.batch_sizes[sched]
+        print(f"Passed argument batch_size = auto:{self.batch_schedule}. Detecting largest batch size")
+        self.batch_sizes[sched] = self._detect_batch_size(n_reordered_requests, pos)
+        print(f"Determined largest batch size: {self.batch_sizes[sched]}")
+        return self.batch_sizes[sched]
+
+    def _loglikelihood_tokens(
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:600]: What does this do?

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,946 @@
...
+                    pbar.update(1)
+
+        pbar.close()
+
+        return re_ord.get_original(res)
+
+    def generate_until(self, requests: list[lm_eval.api.instance.Instance], disable_tqdm: bool = False) -> list[str]:
+        res = []
+
+        def _collate(req: tuple[str, dict]):
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:805]: lambda

File: fast_llm/engine/evaluation/lm_eval/utils.py [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+    assert not args.model_args  # default empty string
+    assert args.batch_size == 1  # default value of 1
+    assert args.max_batch_size is None
+    assert args.device is None
+    # if args.wandb_args:
+    #     wandb_args_dict = simple_parse_args_string(args.wandb_args)
+    #     wandb_config_args_dict = simple_parse_args_string(args.wandb_config_args)
+    #     wandb_logger = WandbLogger(wandb_args_dict, wandb_config_args_dict)
+
+    # TODO: change logging levels from fast_llm to lm_eval and then back?
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/utils.py:70]: Why would we want to do that?

File: fast_llm/engine/evaluation/lm_eval/utils.py [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+
+    # NOTE: all this args are set by fast_llm on the model directly or not used here
+    assert not args.wandb_args  # default empty string
+    assert not args.wandb_config_args  # default empty string
+    assert args.model == "hf"  # default value of 'hf'
+    assert not args.model_args  # default empty string
+    assert args.batch_size == 1  # default value of 1
+    assert args.max_batch_size is None
+    assert args.device is None
+    # if args.wandb_args:
```
Comment (jlamypoirier) [inline fast_llm/engine/evaluation/lm_eval/utils.py:65]: Unnecessary commented code

File: fast_llm/engine/training/wandb.py [hunk 15]
```diff
@@ -44,12 +44,12 @@ def __init__(self, config: WandbConfig, run: Run, experiment_config: Config):
...
             self._wandb = None
 
-    def log_metrics(self, completed_steps: int, metrics: dict[str, dict[str, float | int]]) -> None:
+    def log_metrics(self, completed_steps: int, metrics: dict[str, dict[str, float | int]], commit: bool) -> None:
         # Note: metrics modified in-place
         if self._wandb is not None:
             import wandb
 
-            wandb.log(metrics, step=completed_steps)  # noqa
+            wandb.log(metrics, step=completed_steps, commit=commit)  # noqa
```
Comment (jlamypoirier) [inline fast_llm/engine/training/wandb.py:52]: Why do we want this?

Comment (bigximik) [reply fast_llm/engine/training/wandb.py:52]: This approach allows the code to log multiple times for the same step. When the final log for that step is recorded, all previous logs for that step become immediately visible in Weights & Biases (wandb). There's no longer a need to wait for the next step to start logging for the previous step's logs to appear. Previously, logs for a step would remain hidden until logging began for the next step.

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 4 above)
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:312]: They differ by the presence of `attention_mask` as one of the arguments—this is how it was implemented in the Hugging Face (HF) wrapper. I'm not sure why `attention_mask=None` isn't just specified directly. I checked the latest version of `generate`, and it does use `None` as the default, but earlier versions may have relied on the explicit presence of `attention_mask` in the argument list instead.

Because of this, I would prefer to leave it as the HF team originally wrote it.

What we can do, however, is consolidate the arguments by using a `kwargs` dictionary in a single call instead.

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 5 above)
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:344]: the same as above

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 16]
```diff
@@ -0,0 +1,946 @@
...
+            # Must not be called with continue_generate false on one process
+            assert continue_generate
+            return self._model_invoke_inner(
+                input_ids, attention_mask, labels, max_length, stop, generate, **generation_kwargs
+            )
+
+        rank = self.group.rank()
+        assert rank == 0
+
+        if continue_generate:
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:122]: Yeah, I thought of this. However, I can't be 100% sure in which order `lm_eval` will call tasks, and they also have their own data parallel implementation using `accelerate`, which is not compatible with our data parallel implementation.

So, after reviewing the `lm_eval` code together with @tscholak, we decided to adopt this approach instead—where `lm_eval` runs only on the main rank and the data is distributed across ranks.

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 5 above)
Comment (jlamypoirier) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py]: ee other coment

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 8 above)
Comment (jlamypoirier) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:369]: This implementation doesn't make sense for us though. From what I gather it's really just `return self._max_length`

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py [hunk 17]
```diff
@@ -0,0 +1,946 @@
...
+        self._rank = 0
+        self._world_size = 1
+
+        self._distributed: Distributed = model._inference_runner._fast_llm_model.distributed
+        dist_config: DistributedConfig = self._distributed.config
+        # get batch_data_parallel group leaders
+        if dist_config.sequence_data_rank == 0 and dist_config.pipeline_rank == 0 and dist_config.tensor_rank == 0:
+            self.group = self._distributed.batch_data_group
+        else:
+            self.group = torch.distributed.GroupMember.NON_GROUP_MEMBER
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py:45]: we need to distinguish between None and NON_GROUP_MEMBER as None can not run worker_model_invoke but it can be run on NON_GROUP_MEMBER just waiting.

File: fast_llm/cli.py [hunk 18]
```diff
@@ -7,6 +8,16 @@
 from fast_llm.engine.config_utils.run import log_main_rank
 from fast_llm.engine.config_utils.runnable import RunnableConfig
 
+# This must be set before importing numexpr,
+# because by default, the maximum number of threads is 64.
+# On systems with more cores, numexpr logs an error and
+# ignores the thread setting if it exceeds the limit.
+if "NUMEXPR_MAX_THREADS" not in os.environ:
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/cli.py:16]: We do not have `Run.set_global_variables` — it would need to be called before any other imports.  
In fact, all additions or modifications of environment variables in the current code are unsafe.  
At any point, a library could import another module that depends on an environment variable we modify later, which would have no effect.

Therefore, I would say that any environment variable manipulation must be done at the entry point, before any library imports, regardless of whether it's considered clean or not.

File: fast_llm/engine/evaluation/evaluator.py [hunk 19]
```diff
@@ -246,6 +251,104 @@ def _get_data_iterator(
...
+        data: Data,
+        phase: PhaseType,
+    ) -> None:
+        os.environ["HF_TOKEN"] = pathlib.Path(os.environ["HUGGINGFACE_API_KEY_PATH"]).open("r").read().strip()
+
+        from fast_llm.engine.evaluation.lm_eval.fast_llm_wrapper import FastLLMLmEvalWrapper
+
+        super().setup(distributed, run, multi_stage, runner, data, phase)
+
+        self._hf_model = self._multi_stage.config_class.get_huggingface_model_for_causal_lm_class()(
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/evaluation/evaluator.py]: I have added to HF wrapper class

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 4 above)
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py]: Made it one call

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 9 above)
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py]: I went through the lm_eval code once more and kept only the properties that I found are actually used, or that overrids from the base class—even if they're not used externally by lm_eval.

File: fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py (hunk 10 above)
Comment (bigximik) [reply fast_llm/engine/evaluation/lm_eval/fast_llm_wrapper.py]: yeh, tuple(a) +tuple(b) is even faster

File: tests/models/test_lm_eval_simple.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+        f"data.tokenizer.path={tokenizer_path}",
+        f"model.base_model.vocab_size={tokenizer.vocab_size}",
+        "training.evaluators.evaluation_test.interval=1",
+        "training.evaluators.evaluation_test.evaluator.type=lm_eval",
+        "training.evaluators.evaluation_test.evaluator.cli_args="
+        f'["--tasks","gsm8k,xnli_en,wikitext","--output_path","{str(base_path / "lm_eval")}","--limit","10"]',
+    ]
+
+
+@pytest.mark.extra_slow
```
Comment (jlamypoirier) [inline tests/models/test_lm_eval_simple.py:34]: How long does this take? It would be worrying not to have any tests other than extra-slow.

File: tests/utils/utils.py [hunk 21]
```diff
@@ -11,6 +11,15 @@
 
 requires_cuda = pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA is not available")
 
+try:
```
[the remark this answers is not in the document]
Comment (bigximik) [reply tests/utils/utils.py:14]: as you suggested i moved it to test file itself

File: tests/models/test_lm_eval.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+
+
+@pytest.fixture(scope="module")
+def tokenizer_path():
+    download_santacoder_tokenizer()
+    return TOKENIZER_PATH
+
+
+@pytest.fixture(scope="function")
+def get_lm_eval_config(tokenizer_path, monkeypatch):
```
Comment (jlamypoirier) [inline tests/models/test_lm_eval.py:22]: I managed to reduce test time by trimming bloat in lm_eval, reducing evaluation size and restricting task to wikitest (others are much slower). Should still be enough since we're only testing fast-llm's wrapper, not lm_eval itself. It's now below 10 seconds so we no longer need to mark as extra_slow.

**Review by jlamypoirier — COMMENTED** [review]
Should be ready to merge once remaining comment is addressed and tests pass

File: tests/models/test_lm_eval.py (hunk 22 above)
Comment (bigximik) [reply tests/models/test_lm_eval.py:22]: cool, thanks

File: setup.cfg [hunk 23]
```diff
@@ -51,6 +51,11 @@ HUGGINGFACE =
 #   MAMBA_FORCE_BUILD=TRUE CAUSAL_CONV1D_FORCE_BUILD=TRUE CAUSAL_CONV1D_SKIP_CUDA_BUILD=TRUE pip install -e ".[CORE,SSM]" --no-build-isolation
 SSM =
     mamba_ssm[causal-conv1d]==2.2.4
+    cartesia_pytorch>=0.0.2
+
+GENERATION =
+    lm_eval>=0.4.9
```
[the remark this answers is not in the document]
Comment (bigximik) [reply setup.cfg:57]: Maybe the section should be named 'evaluations', since it's not specifically for generation?

File: tests/models/test_lm_eval.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+
+@requires_cuda
+@pytest.mark.depends_on(on=["test_lm_eval_in_training[{model_testing_config}]"])
+@pytest.mark.model_testing_group(ModelTestingGroup.generate)
+def test_lm_eval_evaluation(run_test_script_for_all_models, run_test_script_base_path, get_lm_eval_config):
+    run_test_script_for_all_models(
+        get_lm_eval_config(run_test_script_base_path / "test_lm_eval_in_training"),
+        compare="test_lm_eval_in_training",
+        prepare_fn=_prepare_resume_fn,
```
[the remark this answers is not in the document]
Comment (bigximik) [reply tests/models/test_lm_eval.py]: Okay, I created a function that simply copies the training run with a checkpoint, but the evaluation still starts from the training config. I haven’t tested it starting from a pretrained checkpoint yet — maybe we should leave that for another PR?

[the remark this answers is not in the document]
Comment (jlamypoirier) [reply tests/models/test_lm_eval.py]: Isn't this a 5 minute thing? Just need to add 2 arguments (pretrained path and format)...

[the remark this answers is not in the document]
Comment (bigximik) [reply tests/models/test_lm_eval.py]: Mainly because, as it stands now, this is already a valid test case — evaluating the most recent checkpoint right after training, where evaluation is appended after the training run.

For evaluating pretrained models as a separate step, we’ll need to design a dedicated evaluation config, especially since there are requests to evaluate multiple checkpoints at once.

That’s why I propose handling this in a separate PR — since it involves the design and implementation of the evaluation config.

[the remark this answers is not in the document]
Comment (bigximik) [reply tests/models/test_lm_eval.py]: But ok, I will add an additional test now for a training config that starts from a pretrained model which hasn't been trained yet. So we will have 2 evaluation tests one from last checkpoint of the trained model and one from pre-trained model.

**Review by jlamypoirier — APPROVED** [review]
